hw-36.ts

```diff
@@ -0,0 +1,165 @@
+        director: "Chris Columbus",
+        writer: "J.K. Rowling, Steve Kloves",
+        actors: ["Daniel Radcliffe", "Rupert Grint", "Richard Harris"],
+        plot: "An orphaned boy enrolls in a school of wizardry, where he learns the truth about himself, his family and the terrible evil that haunts the magical world.",
+        country: "United Kingdom, United States",
+        poster: "https://m.media-amazon.com/images/M/MV5BNjQ3NWNlNmQtMTE5ZS00MDdmLTlkZjUtZTBlM2UxMGFiMTU3XkEyXkFqcGdeQXVyNjUwNzk3NDc@._V1_SX300.jpg",
+        imdbRating: 7.6,
+        imdbVotes: 684604,
+        boxOffice: "$318,087,620",
+        production: "1492 Pictures, Heyday Films, Warner Brothers",
+    }
+
+]
+
+// 1. Все жанры фильмов без повторения
+const getGenres = (array: Movie[]) => {
+    const arrayOfAllGenres = array.map(movie => movie.genre)
+    const allGenres = []
+    const genres = allGenres.concat(...arrayOfAllGenres)
+    const resultArray = Array.from(new Set(genres))
+    return resultArray
+}
+console.log(getGenres(movies));
+
+// 2. Массив актёров без повторения
+const getActors = (array: Movie[]) => {
+    const arrayOfAllActors = array.map(movie => movie.actors)
+    const allActors = []
+    const actors = allActors.concat(...arrayOfAllActors)
+    const resultArray = Array.from(new Set(actors))
+    return resultArray
+}
+console.log(getActors(movies));
+
+// 3. Сортировка по рейтингу по убыванию
+const sortByImdbRating = (array: Movie[]) => {
+    const sortedArray = array.sort((a, b) => b.imdbRating - a.imdbRating)
+    return sortedArray
+}
+console.log(sortByImdbRating(movies));
+
+// 4. Новый массив, где объекты фильмов состояни из полей id, title, released, plot
+const newMoviesArray = (array: Movie[]) => array.map(movie => ({id: movie.id, title: movie.title, released: movie.released, plot: movie.plot}))
+console.log(newMoviesArray(movies));
+
+// 5. Функция принимает массив и число. Число = году выхода фильма
+const filteredMoviesByYear = (array: Movie[], num: number) => array.filter(movie => movie.year === num)
+console.log(filteredMoviesByYear(movies, 2009));
+
+// 6. Функция принимает массив и строку. Строка входит в название фильма
+const filteredMoviesByName = (array: Movie[], word: string) => array.filter(movie => movie.title.toLowerCase().includes(word.toLowerCase()) === true)
+console.log(filteredMoviesByName(movies, 'haRRy'));
+
+// 7. Функция принимает массив и строку. Строка входит в название фильма или сюжет
+const filteredMoviesByWords = (array: Movie[], word: string) => array.filter(movie => 
+    (movie.title.toLowerCase().includes(word.toLowerCase()) === true) || (movie.plot.toLowerCase().includes(word.toLowerCase()) === true))
+console.log(filteredMoviesByWords(movies, 'a cocky Pilot'));
+
+// 8. Функция принимает три параметра 1.Массив фильмов, 2.Строка (ключ), 3.Строка/число (значение)
+const researchMovie = (array: Movie[], key: string, data: string | number) => array.filter(movie => movie[key] === data)
```

Comment:
key: keyof Movie

Reply:
I can't understand this. Why should i write keyof Movie if I know, that it is a string. Should i write this way for every parameter that I put to function? Why it doesn't work here: "data: keyof Movie" then?
And where do you see git ussues? What should I do?

Reply:
because keyof Movie is more specific type
it mean that it's not any string, but one of Movie keys
data - is any string or number, and doesn't connect to movies keys